### PR TITLE
Add support for Contact tokens in welcome email using TokenProcessor

### PR DESCRIPTION
**Pull request by masetto**
Overview
----------------------------------------
Add the possibility of using contact tokens in the welcome email

Before
----------------------------------------
In the welcome email contact tokens can not be used.

After
----------------------------------------
In the welcome email contact tokens can be used. For example:

```
Dear {contact.first_name} {contact.last_name},
you are welcome to the {welcome.group}.

To view or update your communication preferences: {contact.comm_pref_supporter_link}
```

Technical Details
----------------------------------------
Add new token processor.
We continue to uses `replaceWelcomeTokens` because  I don't think it is possible to replace them using token processor.

### Note
_This is a partial pull request. I would like to understand if it is OK the way I did it. Then it would be replicated for`Resubscribe`, `Subscribe`, `Unsubscribe` messages._

### CODE REVIEW

File: CRM/Mailing/Event/BAO/Confirm.php [hunk 1]
```diff
@@ -108,17 +108,22 @@ public static function confirm($contact_id, $subscribe_id, $hash) {
...
-
-    $html = CRM_Utils_Token::replaceDomainTokens($html, $domain, TRUE, $tokens['html']);
     $html = CRM_Utils_Token::replaceWelcomeTokens($html, $group->title, TRUE);
-
-    $text = CRM_Utils_Token::replaceDomainTokens($text, $domain, FALSE, $tokens['text']);
     $text = CRM_Utils_Token::replaceWelcomeTokens($text, $group->title, FALSE);
 
+    $tokenProcessor = new \Civi\Token\TokenProcessor(\Civi::dispatcher(), [
+      'controller' => __CLASS__,
+      'smarty' => FALSE,
```
Comment (eileenmcnaughton) [inline CRM/Mailing/Event/BAO/Confirm.php:116]: Also specifiy `'schema' => ['contactId'] ` here as best practice

**Comment by eileenmcnaughton** [comment]
Yes this looks great - ideally the `welcome` tokens WOULD be moved to the token processor but that is a slightly bigger challenge

I think the way we would do it is 
- add a new class  - https://github.com/civicrm/civicrm-core/compare/master...eileenmcnaughton:civicrm-core:welcome?expand=1
- Replace `{welcome.group}` in `civicrm_mailing_component ` with `{group.title}` (in an upgrade script) 
- add 'groupId' to the schema & row in this PR

**Comment by masetto** [comment]
@eileenmcnaughton I think it is too complex for me  to do what you are asking ....maybe we do this in another PR? 

In welcome email I'd also like to replace action tokens using `CRM_Utils_Token::replaceActionTokens`: would allow us to write for example:
```
You can <a href="{action.unsubscribeUrl}" >unsubscribe from this mailing</a> at any time.
```
But I would have to use this instruction `CRM_Mailing_BAO_Mailing::getVerpAndUrls($job, $queue_id, $eq->hash, $eq->email)` as it is done in `CRM_Mailing_Event_BAO_Resubscribe` class, but I don't have the "job" here...

I took a look at "Resubscribe.php" and there in addition  it is called `replaceMailingTokens`: token process have a specific schema for that?

**Comment by eileenmcnaughton** [comment]
@masetto - yes I think it's fine for the scope of this PR to just be replacing those `CRM_Utils_Token::replaceDomainTokens` functions & adding in the `ContactToken` support. Ideallly it would have a test but I think these functions are pretty tricky to write tests on so we could probably let that slip

**Comment by masetto** [comment]
@eileenmcnaughton I would have finished and I just replace `CRM_Utils_Token::replaceDomainTokens`. Let me know if I can squash commits.

**Comment by masetto** [comment]
@eileenmcnaughton, my changes cause a regression :cry: : in `Subscribe.php` and `Confirm.php` I removed this code:

```
    $bao = new CRM_Mailing_BAO_Mailing();
    $bao->body_text = $text;
    $bao->body_html = $html;
    $tokens = $bao->getTokens();
```
since `$tokens` was no longer used.
But the function `getTokens` calls the hook `alterMailContent` with this stack trace (for subscription): 

```
#0  CRM_Mailing_BAO_Mailing->getTemplates() called at [CRM/Mailing/BAO/Mailing.php:784]
#1  CRM_Mailing_BAO_Mailing->_getTokens(html) called at [CRM/Mailing/BAO/Mailing.php:730]
#2  CRM_Mailing_BAO_Mailing->getTokens() called at [CRM/Mailing/Event/BAO/Subscribe.php:226]
#3  CRM_Mailing_Event_BAO_Subscribe->send_confirm_request called at [CRM/Mailing/Event/BAO/Subscribe.php:357]
#4  CRM_Mailing_Event_BAO_Subscribe::commonSubscribe called at [CRM/Mailing/Form/Subscribe.php:154]
```
It does not seem appropriate to re-introduce this code, because anyone would wonder what it is for. 
What do you suggest we do?

I have not checked but maybe 'getTokens' does something else...


and now this hook is not called.

**Comment by eileenmcnaughton** [comment]
@masetto is the regression that `alterMessageContent` is now skipped? It seems like ideally we would still call `getTemplates` to retrieve the html, text & subject - but bypass the `getTokens` part?

**Comment by masetto** [comment]
@eileenmcnaughton I had a mistake: the `alterMessageContent` hook was called by `getTokens` function, but the content of templates was never used.
So I (re)declare `CRM_Mailing_BAO_Mailing` object, then I call `getTemplates` function and I use `$templates['html']` and `$templates['text']` instead of `$html` and `$text`. Now  `alterMessageContent` hook is called and the message content is really the "altered content".

**Comment by eileenmcnaughton** [comment]
@masetto hmm - so the hook wasn't working before - is it an issue for you for it to work? Otherwise if it wasn't working I'd just let it drop it  (since we still have more cleanup to get that code off the other tokens in that lot & it would be easier not to work around that hook if it never worked in the first place)

**Comment by masetto** [comment]
@eileenmcnaughton Now the hook works and I think it's fine :stuck_out_tongue:
I realise that it solves an issue unrelated to this PR, but I would need it: I am using an extension that adds a header and footer to all system emails and automated messages are not being intercepted. Now they do.

But if you want I can do another PR after you merge this one.

**Comment by eileenmcnaughton** [comment]
@masetto that's fine - it can stay as it is - are you happy with this now / ready to remove the WIP

**Comment by eileenmcnaughton** [comment]
Looks good - great improvement

**Comment by eileenmcnaughton** [comment]
@masetto I took a look at the last place with the `replaceDomainTokens` but then as I dug into it I started to think that this actually causes a regression - the templates loaded by ` $templates = $bao->getTemplates();` are actually different components I think.... it's loading the header & footer rather than the relevant component

https://github.com/civicrm/civicrm-core/pull/24446

**Comment by eileenmcnaughton** [comment]
We really should add an api for this stuff - there is the `sendTemplate` function but it is really hard to understand how to use it & I think there is another 'light weight' version but we just need an api with sensible params not a bunch of internal functions

**Comment by eileenmcnaughton** [comment]
OK - I updated ^^ with what I *think* will work to use the right template AND call the hook....

**Comment by masetto** [comment]
Oh, great! Now we could also adapt these classes to use `CRM_Core_BAO_MessageTemplate` ...

**Pull request merged**